### ISSUE TITLE
Files in /data are not available

### ISSUE THREAD

**Issue by greinbold** [shaarli/Shaarli#1667]
Hi

I seen that many files in data was unavailable because, in accordance with the documentation (https://shaarli.readthedocs.io/en/master/Server-configuration/#apache), we have to set in the Apache config:

```
    <FilesMatch ".*\.(?!(ico|css|js|gif|jpe?g|png|ttf|oet|woff2?)$)[^\.]*$">
        Require all denied
    </FilesMatch>

    <Files "index.php">
        Require all granted
    </Files>

    <FilesMatch "\.(?:ico|css|js|gif|jpe?g|png|ttf|oet|woff2)$">
        # allow client-side caching of static files
        Header set Cache-Control "max-age=2628000, public, must-revalidate, proxy-revalidate"
    </FilesMatch>
```

For example, this setting break custom_css. Maybe other plugins?

When I try to access to the custom.css, I always got a Forbidden access error because the custom.css file lives in the data folder.

I'm behind a HaProxy with a very simple and standard config like:

```
frontend http
    bind :443 ssl crt /etc/ssl/haproxy/
    bind :80
    acl shaarli ...
    redirect scheme https if !{ ssl_fc } shaarli

backend shaarli
    option forwardfor
    option httpchk
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-request set-header X-Forwarded-Port %[dst_port]
    server apache x.x.x.x:xx
```

In my Apache config, I also have enabled the rewrite and header modules.

My Shaarli instance https://shaarli.nosheep.fr works well, but I do not have access to the files in data folder.

Cheers

**Comment by greinbold** [comment]
Just a simple question: why we have to denied access to the css and image files in data folder?

**Comment by ArthurHoaro** [comment]
Hi,

The Apache documentation part that you quoted doesn't contain any directive specific to `data/` folder and should be able to load any static file.

However looking at `.htaccess` in data folder, that probably where your issue comes from: it only allows access to `user.css`. This restriction was added to avoid leaking any sensitive information from the data folder, but it probably could be improved to allow access to any static file.

**Comment by immanuelfodor** [comment]
Allowing `*.css` and `*.js` access maybe a good idea. 

By the way, it's a security best practice to move sensitive files outside of the web root, and not only depend on a htaccess file that might get misconfigured one time. Quick examples:
- Nextcloud: https://docs.nextcloud.com/server/latest/admin_manual/installation/installation_wizard.html?highlight=data#data-directory-location-label
- CVE: https://cwe.mitre.org/data/definitions/219.html
- Random SO question: https://stackoverflow.com/questions/3034474/storing-script-files-outside-web-root

**Comment by ArthurHoaro** [comment]
You are right, and it should probably have been done a while ago. However I don't really want to break every user configuration, Docker image, 3rd party theme and plugins, etc. with this change. 

The documentation provides a complete guide to configure web servers securely, we should rely on that. As for `.htaccess` files, I see them as a bonus for users who use Apache without any custom configuration, but they are kind of legacy to be fair.

**Comment by nodiscc** [comment]
- [ ] Allowing ` data/*.css` and `data/*.js` access maybe a good idea.

**Comment by nodiscc** [comment]
I've been trying to reproduce this problem. `data/user.css` is successfully loaded, as it is allowed by `.htaccess`. Attempting to load other `data/*.css` files will be denied, as per `.htaccess`. But I do not see any valid reason to attempt loading CSS files other than `user.css`

> For example, this setting break custom_css.

This is because this plugin writes its CSS rules to `data//custom_css.css` (https://github.com/immanuelfodor/shaarli-custom-css/blob/master/custom_css/custom_css.php#L60). The proper solution here is to fix the plugin to write to `data/user.css` instead.

Closing, feel free to reopen with more information if needed.

**Comment by OrinZ** [comment]
>  The proper solution here is to fix the plugin to write to `data/user.css` instead.

This worked for me! I [forked the repo ](https://github.com/OrinZ/shaarli-custom-css/tree/master)to implement the changes, very simple indeed.
